src/CodeGen_Hexagon.cpp

```diff
@@ -175,6 +157,69 @@ Stmt sloppy_unpredicate_loads(Stmt s) {
     return SloppyUnpredicateLoads().mutate(s);
 }
 
+class InjectHVXLocks : public IRMutator2 {
+public:
+    InjectHVXLocks(const Target &t) : target(t) {}
+    bool uses_hvx = false;
+private:
+    using IRMutator2::visit;
+
+    Stmt visit(const For *op) {
+        if (op->for_type == ForType::Parallel) {
+            bool old_uses_hvx = uses_hvx;
+            uses_hvx = false;
+            Stmt body = mutate(op->body);
+            if (uses_hvx) {
+                body = acquire_hvx_context(body, target);
+            }
+            uses_hvx = old_uses_hvx;
+            Stmt new_for = For::make(op->name, op->min, op->extent, op->for_type,
+                             op->device_api, body);
+            if (old_uses_hvx) {
```

Comment:
I'm realizing that this won't quite work, if code after a parallel for loop uses HVX, a qurt_hvx_lock call will be added to the code before the loop, but this code won't get injected as necessary.

I can think of one not-so-pretty solution:

- Prior to the parallel for loop, always inject the following (this is Halide IR, `uses_hvx_var` is a Variable of type bool with name "uses_hvx"):
    if (uses_hvx_var) {
        halide_qurt_hvx_unlock();
    }

- After the loop, insert:
    if (uses_hvx_var) {
        halide_qurt_hvx_lock();
    }

- When exiting an HVX locked scope, i.e. any time you *might* call `acquire_hvx_context`, do:
    body = substitute(body, "uses_hvx", uses_hvx);

The idea is that you don't know whether you need to unlock/relock HVX until after the enclosing acquire_hvx_context call is injected. This strategy allows you to defer that decision until you actually call (or not) acquire_hvx_context.

The if will always be a constant, and so it will simplify away.

Reply:
Ok I see what you mean. You mean that if we have code that is like so.

```
serial_for() {
    parallel_for() {
        vector_code
  }
  vector_code
}
```
Then we'll end up with no lock getting injected after the parallel loop because this logic relies on there being vector code in serial_for before parallel_for

Thanks for taking the time to catch this for me. Let me digest and understand the entirety of your approach and come up with a solution.